storage/src/mmr/bitmap.rs

```diff
@@ -0,0 +1,290 @@
+    /// chunk is given by `self.next_bit`.
+    bitmap: Vec<u8>,
+
+    /// The position within the last chunk of the bitmap where the next bit is to be appended.
+    ///
+    /// Invariant: This value is always in the range [0, DIGEST_SIZE * 8).
+    next_bit: u64,
+
+    /// A Merkle tree with each leaf representing DIGEST_SIZE*8 bits of the bitmap.
+    ///
+    /// When a chunk of DIGEST_SIZE*8 bits is accumulated by the bitmap, it is added to this tree.
+    /// Because leaf elements can be updated when bits in the bitmap are flipped, this tree, while
+    /// based on an MMR structure, is not an MMR but a Merkle tree.  The MMR structure results in
+    /// reduced update overhead for elements being appended or updated near the tip compared to a
+    /// more typical balanced Merkle tree.
+    mmr: Mmr<H>,
+}
+
+impl<H: CHasher> Default for Bitmap<H> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<H: CHasher> Bitmap<H> {
+    const CHUNK_USIZE: usize = H::Digest::SERIALIZED_LEN;
+    const CHUNK_SIZE: u64 = H::Digest::SERIALIZED_LEN as u64;
+
+    /// Return a new empty bitmap.
+    pub fn new() -> Self {
+        Bitmap {
+            bitmap: vec![0u8; H::Digest::SERIALIZED_LEN],
+            next_bit: 0,
+            mmr: Mmr::new(),
+        }
+    }
+
+    /// Return the number of bits currently stored in the bitmap.
+    pub fn bit_count(&self) -> u64 {
+        self.bitmap.len() as u64 * 8 - (Self::CHUNK_SIZE * 8 - self.next_bit)
+    }
+
+    /// Commit the last chunk of the bitmap to the Merkle tree and initialize the next chunk.
+    fn commit_last_chunk(&mut self, hasher: &mut H) {
+        let len = self.bitmap.len();
+        let chunk = H::Digest::try_from(&self.bitmap[len - Self::CHUNK_USIZE..len]).unwrap();
+        self.mmr.add(hasher, &chunk);
+        self.next_bit = 0;
+        self.bitmap.extend(vec![0u8; Self::CHUNK_USIZE]);
+    }
+
+    /// Efficiently add a digest-sized chunk of bits to the bitmap. Assumes we are at a chunk
+    /// boundary (that is, `self.next_bit` is 0) and panics otherwise.
+    pub fn append_chunk_unchecked(&mut self, hasher: &mut H, chunk: &H::Digest) {
+        if self.next_bit != 0 {
+            panic!("cannot add chunk when not chunk aligned");
+        }
+
+        let len = self.bitmap.len();
+        let _ = &self.bitmap[len - Self::CHUNK_USIZE..len].copy_from_slice(chunk.as_ref());
+        self.commit_last_chunk(hasher);
+    }
+
+    /// Efficiently add a byte's worth of bits to the bitmap. Assumes self.next_bit is currently
+    /// byte aligned, and panics otherwise.
+    pub fn append_byte_unchecked(&mut self, hasher: &mut H, byte: u8) {
+        if self.next_bit % 8 != 0 {
+            panic!("cannot add byte when not byte aligned");
+        }
+
+        let len = self.bitmap.len() - Self::CHUNK_USIZE;
+        self.bitmap[len + self.next_bit as usize / 8] = byte;
```

Comment:
feels like you should be using `bit_count()` as a helper function rather than re-calculating it slightly differently here and in `append()`.

E.g.:
`self.bitmap[self.bit_count() as usize / 8] = byte;`
Or if you have a helper
`self.bitmap[bit_to_byte_offset(self.bit_count())] = byte;`

Reply:
After the above changes I don't think there's any new place where using bit_count() might be an improvement.